src/FirewallManager.cs

```diff
@@ -0,0 +1,1096 @@
+﻿
+/*
+Supports disabling of default firewalls and opening up ports that can be accessed remotely.
+
+.NET port of landaire's hard work: https://github.com/exploits-forsale/solstice/blob/main/crates/solstice_daemon/src/firewall.rs
+
+Example API usage:
+```
+   FirewallManager.DisableFirewalls();
+   FirewallManager.AllowPortThroughFirewall("Debugger", 23946);
+   FirewallManager.AllowPortThroughFirewall("SSH", 22);
+```
+
+Example E2E usage:
+1. Copy .NET to `D:\XBOX\dotnet\`
+2. Copy this DLL (`DurangoInteropDotnet.dll`) to `D:\XBOX\payloads\`
+3. Copy `msbuild_tasks/disable_firewall.xml` to `D:\XBOX\payloads\disable_firewall.xml`
+4. Execute the following commands:
+```
+    set DOTNET_CLI_TELEMETRY_OPTOUT=1
+    set DOTNET_EnableWriteXorExecute=0
+    D:\XBOX\dotnet\dotnet.exe msbuild "D:\XBOX\payloads\disable_firewall.xml"
+```
+*/
+using System.Runtime.InteropServices.ComTypes;
+
+namespace DurangoInteropDotnet {
+
+    public static class FirewallManager {
+
+        private static readonly string FWPM_DISPLAY_NAME = "XOR";
+        private static readonly string FWPM_DISPLAY_DESCRIPTION = "XOR FWPM Provider";
+        private static readonly Guid FWPM_PROVIDER_GUID = new Guid(0xabad1dea, 0x4141, 0x4141, 0x0, 0x0, 0x0c, 0x0f, 0xfe, 0xe0, 0x00, 0x00);
+        private static readonly Guid CLSID_HNETCFGFWPOLICY2 = new Guid("e2b3c97f-6ae1-41ac-817a-f6f92166d7dd");
+        private static readonly Guid CLSID_INETFWPOLICY2 = new Guid("98325047-C671-4174-8D81-DEFCD3F03186");
+
+        public static void DisableFirewalls() {
+
+            uint result = NativeBridge.CoInitializeEx(IntPtr.Zero, NativeBridge.COINIT.COINIT_MULTITHREADED);
+            if (result != 0 && result != 1) { // 0 = S_OK, 1 = S_FALSE
```

Comment:
This looks a bit odd?

Reply:
From Landaire's rust code:
```
        CoInitializeEx(None, COINIT_MULTITHREADED)
            .ok()
            .context("CoInitializeEx")?;
```

As for the result, [MSDN states](https://learn.microsoft.com/en-us/windows/win32/api/combaseapi/nf-combaseapi-coinitializeex#return-value) that the following for `CoInitializeEx()`:

| Return Code | Description |
| --- | --- |
| S_OK | The COM library was initialized successfully on this thread. |
| S_FALSE | The COM library is already initialized on this thread. |

I assumed both of these scenarios are nominal for our use case.  Happy to be wrong.

Reply:
Thanks for the wakeup call, you are absolutely right!